### PR TITLE
Fix postgres pkey map in Qt5 (fixes #15223)

### PR DESCRIPTION
**Pull request by wonder-sk**
Switching from QVariant to QVariantList solves the underlying Qt5 issue:
- comparison of QVariantList objects works fine
- comparison of QVariantList objects wrapped in QVariant does not work

For Qt4 we have our own QVariant comparison operator in postgres provider that handles QVariantList correctly, but in Qt5 the QVariant comparison is given.

The extra wrapping of QVariantList into another QVariant seems unnecessary anyway, so we may as well save a tiny bit of memory and cpu

Needs someone to test properly on qt4/qt5... maybe @3nids ? :-)

cc @nyalldawson

### CODE REVIEW

**Comment by wonder-sk** [comment]
Strange that qt5 build on travis fails, it builds fine for me with qt5.7

**Comment by wonder-sk** [comment]
Looks like operator< for QList has been added only in Qt 5.6

File: src/providers/postgres/qgspostgresprovider.cpp [hunk 1]
```diff
@@ -403,6 +403,17 @@ static bool operator<( const QVariant &a, const QVariant &b )
 }
 #endif
 
+#if QT_VERSION >= 0x050000 && QT_VERSION < 0x050600
```
Comment (nyalldawson) [inline src/providers/postgres/qgspostgresprovider.cpp]: Maybe this should be defined in qgis.h? qgsVariantLessThan could benefit from using it too.

Comment (wonder-sk) [inline src/providers/postgres/qgspostgresprovider.cpp]: Hmm... maybe yes..

btw. looking at the code, just above the "less than" operator for QVariant should probably use qgsVariantLessThan?

Comment (nyalldawson) [inline src/providers/postgres/qgspostgresprovider.cpp]: Where are you referring to?

**Comment by 3nids** [comment]
I'll do some testing on Qt5!

File: src/providers/postgres/qgspostgresprovider.cpp [hunk 2]
```diff
@@ -4291,11 +4296,11 @@ void QgsPostgresSharedData::setFeaturesCounted( long count )
 }
 
 
-QgsFeatureId QgsPostgresSharedData::lookupFid( const QVariant &v )
+QgsFeatureId QgsPostgresSharedData::lookupFid( const QVariantList& v )
 {
   QMutexLocker locker( &mMutex );
 
-  QMap<QVariant, QgsFeatureId>::const_iterator it = mKeyToFid.constFind( v );
+  QMap<QVariantList, QgsFeatureId>::const_iterator it = mKeyToFid.constFind( v );
```
Comment (3nids) [inline src/providers/postgres/qgspostgresprovider.cpp]: Wouldn't it be worth to swtich to QHash too?
From what I read, it is supposed to be more efficient with 20 and more objects.

Comment (nyalldawson) [inline src/providers/postgres/qgspostgresprovider.cpp]: Yeah, I've been concerned about the performance impact of forcing everything through this map too.

**Comment by nyalldawson** [comment]
Before we merge this, I'm curious to know why the provider test suite isn't failing as a result. There must be a missing test we need to implement there.

File: src/providers/postgres/qgspostgresprovider.cpp (hunk 1 above)
Comment (wonder-sk) [inline src/providers/postgres/qgspostgresprovider.cpp]: I was referring to this: https://github.com/qgis/QGIS/blob/master/src/providers/postgres/qgspostgresprovider.cpp#L322

File: src/providers/postgres/qgspostgresprovider.cpp (hunk 2 above)
Comment (wonder-sk) [inline src/providers/postgres/qgspostgresprovider.cpp]: QHash would be a better choice, I was just keen to understand the problem and fix it with the least amount of effort / changes ;-) QHash will just require our own qHash implementation for QVariant I suppose.

**Comment by 3nids** [comment]
@nyalldawson There is no test with compound pkey, I am ading this

**Comment by 3nids** [comment]
#3387 for new test

**Comment by nyalldawson** [comment]
Don't we always use the map since e763b41c08b699a2e26a9f3ae114fb8d90fca3e1 ?

**Pull request merged**